engine-src/Game/LambdaHack/Client/MonadClient.hs

```diff
@@ -76,11 +77,11 @@ rndToAction r = do
 rndToActionForget :: MonadClientRead m => Rnd a -> m a
 rndToActionForget r = do
   gen <- getsClient srandom
-  let i = fst $ R.next gen
+  let i = fst $ SM.nextWord32 gen
   time <- getsState stime
-  -- Prevent overflow from @Int64@ to @Int@.
-  let positiveIntSize = finiteBitSize (1 :: Int) - 1
+  -- Prevent overflow from @Int64@ to @Word32@.
+  let positiveIntSize = finiteBitSize (1 :: Word32) - 1
```

Comment:
Here I was measuring how wide `Int` is. If we stick to `Word32`, we can simplify this even more. However, if we sometimes use Word64, this should be reflected here, e.g., via a CPP ifdef.

BTW, I need to add a comment that I do the trimming of time to the width of the random seeds to avoid the bug I had, in which only the bits of time were used that were always zero (unless the game runs for ages) and so time didn't add any (weak) randomness at all.

Reply:
So do I understand that you want deterministically create new generator, perturbed by current time?

I'm not fully understand why you want to perturb using time, and not simply `split` (and record second half back to `m`?)

Perturbing using an `Int` would make sense, and you can get reasonable `Int` from anything using `hashable` (and e.g. https://hackage.haskell.org/package/hashable-time) so you don't need to bit-twiddle yourself.

Reply:
Purturbed by *game time*, which is fully deterministic, though also variable enough to give interesting results, that is, different every player move. OTOH, `split` gives the same results regardless of player moves.

And I don't want to record the changed generator in game state, because this is for UI functions, IIRC, which should not change random seed, which is also used for AI, etc. Otherwise, opening some menus, with UI effects, would affect AI moves and so make reproducing stuff more difficult. OTOH, I think UI could just work with a completely separate random seed, in game session, not in client state. I guess that would be saner.

Yes, hashing would protect me from using only the most significant (and so rarely changed) bits by mistake. I wonder how perfect the hashing of `Word64` into `Int32` is these days and how fast.

Reply:
Even xorring `Word32` halves of `Word64` will make it quite good. One could add an additional mixing step, but if the point is only to let all bits to affect output that's enough (especially as we know that time grows linearly)

Reply:
So, in the end I just added yet another random seed to the state -- to the UI session, to be precise. That's much simpler and keeps UI randomness from interfering the gameplay being deterministic. Thanks for the discussion.